pkg/tnf/handlers/automountservice/automountservice.go

```diff
@@ -0,0 +1,160 @@
+	"github.com/test-network-function/test-network-function/pkg/tnf/reel"
+	"github.com/test-network-function/test-network-function/test-network-function/common"
+)
+
+const (
+	TokenIsTrue  int = 1
+	TokenIsFalse int = 2
+	TokenNotSet  int = 3
+)
+
+// Automountservice is the reel handler struct.
+type Automountservice struct {
+	result         int
+	podCheck       bool
+	namespace      string
+	serviceaccount string
+	podname        string
+	timeout        time.Duration
+	args           []string
+	token          int
+}
+
+const (
+	saRegex = `(?m)"automountServiceAccountToken": (.+)`
+	False   = "false,"
+	True    = "true,"
+)
+
+// NewAutomountservice returns a new Automountservice handler struct.
+func NewAutomountservice(options ...func(*Automountservice)) *Automountservice {
+	as := &Automountservice{
+		timeout: common.DefaultTimeout,
+		result:  tnf.ERROR,
+		token:   TokenNotSet,
+	}
+	for _, o := range options {
+		o(as)
+	}
+	if as.podCheck {
+		as.args = []string{"oc", "-n", as.namespace, "get", "pods", as.podname, "-o", "json", "|", "jq", "-r", ".spec"}
+	} else {
+		as.args = []string{"oc", "-n", as.namespace, "get", "serviceaccounts", as.serviceaccount, "-o", "json"}
+	}
+	return as
+}
+
+// WithNamespace specify the namespace
+func WithNamespace(ns string) func(*Automountservice) {
+	return func(as *Automountservice) {
+		as.namespace = ns
+	}
+}
+
+//
+// WithTimeout specify the timeout of the test
+func WithTimeout(t time.Duration) func(*Automountservice) {
+	return func(as *Automountservice) {
+		as.timeout = t
+	}
+}
+
+// WithPodname specify the podname to test
+func WithPodname(ns string) func(*Automountservice) {
+	return func(as *Automountservice) {
+		as.podname = ns
+		as.podCheck = true
+	}
+}
+
+// WithServiceAccount specify the serviceaccount to check
+func WithServiceAccount(sa string) func(*Automountservice) {
+	return func(as *Automountservice) {
+		as.serviceaccount = sa
+		as.podCheck = false
+	}
+}
+
+// Args returns the initial execution/send command strings for handler Automountservice.
+func (as *Automountservice) Args() []string {
+	return as.args
+}
+
+// GetIdentifier returns the tnf.Test specific identifier.
+func (as *Automountservice) GetIdentifier() identifier.Identifier {
+	return identifier.AutomountServiceIdentifier
+}
+
+// Timeout returns the timeout for the test.
+func (as *Automountservice) Timeout() time.Duration {
+	return as.timeout
+}
+
+// Result returns the test result.
+func (as *Automountservice) Result() int {
+	return as.result
+}
+
+// ReelFirst returns a reel step for handler Automountservice.
+func (as *Automountservice) ReelFirst() *reel.Step {
+	return &reel.Step{
+		Expect:  []string{saRegex},
+		Timeout: as.timeout,
+	}
+}
+
+// ReelMatch parses the Automountservice output and set the test result on match.
+func (as *Automountservice) ReelMatch(_, _, match string) *reel.Step {
+	numExpectedMatches := 2
+	saMatchIdx := 1
+	as.result = tnf.SUCCESS
```

Comment:
I find the handler logic a bit confusing. If I understood it correctly:

- We will always get to ReelMatch, even if the regex is not matched. So we always set as.result = tnf.SUCCESS (which is a bit confusing). However, that will change with https://github.com/test-network-function/test-network-function/pull/496, so I think that PR could break the behavior in this one.

- Then, we re-check the match string against the pattern, and set as.token accordingly only if the match was found.

- Finally, in accesscontrol/suite.go, we check the .token status of both the service account and pod, and that is where we actually decide if the test passes or fails.

Is this correct? If so, it would be great if you could add some comment in the handler code to explain that behavior, otherwise it could be hard to debug in the future. Also, you may need to adjust the pattern to match against if PR 496 is merged, since it will always fail for pods without an explicit automountServiceAccountToken key set.

Reply:
your understanding is correct. I agree that it's bit confusing, but what you described is the logic of most of our tests. 

1- define the command. 
2- run in using generic handlers (via JSON) or using a new handler
3- when running using generic handlers, the status is all that matters and we get the results right away.
4- when choosing new handler, we can return the status or some other data that the test in the suite decides if there's a failure or not. 

*) this specific case here is not intuitive since the result of the Test depends on the results of two separate calls (or handlers)

The part that should be clear is where we decide if there's a failure or not in the suite.go

Reply:
added more comments to make the test clearer in suite.go

Reply:
Thanks for the explanation. It looks much clearer now, with the explanations and the initial match against allRegex in the step.